src/turbomind/kernels/sampling_topp_kernels.h

```diff
@@ -19,135 +19,62 @@
-                             bool*           finished_buf,
-                             float*          cum_log_probs,
-                             float*          output_log_probs,
-                             const T*        log_probs,
-                             float*          sampled_logprobs,
-                             uint32_t*       sampled_indexes,
-                             uint32_t*       sampled_nums,
-                             const int*      id_vals,
-                             int*            offset_buf,
-                             int*            begin_offset_buf,
-                             curandState_t*  curandstate,
-                             const int       batch_size,
-                             const size_t    vocab_size_padded,
-                             const int*      end_ids,
-                             const float     max_top_p,
-                             const float*    top_ps,
-                             cudaStream_t    stream,
-                             cudaDeviceProp* cuda_device_prop,
-                             const bool*     skip_decode);
-
-template<typename T>
-void invokeAddBiasSoftMax(T*           logits,
-                          const T*     bias,
-                          const int*   end_ids,
-                          const bool*  finished,
-                          const int    m,
-                          const int    n_padded,
-                          const int    n,
-                          cudaStream_t stream);
-
-namespace segmented_topp_impl {
-enum DType_t
-{
-    kFLOAT,
-    kHALF,
-    kINT8
-};
-
-template<typename Key_Data_Type_   = float,
-         typename Value_Data_Type_ = int32_t,
-         int BLOCK_THREADS_        = 256,
-         int KEYS_PER_LDG_         = 1>
-struct Segmented_topk_kernel_params {
-    typedef Key_Data_Type_   Key_Data_Type;
-    typedef Value_Data_Type_ Value_Data_Type;
-    enum
+struct BlockPrefixCallbackOp {
+    // Running prefix
+    float running_total;
+    // Constructor
+    __device__ BlockPrefixCallbackOp(float running_total): running_total(running_total) {}
+    // Callback operator to be entered by the first warp of threads in the block.
+    // Thread-0 is responsible for returning a value for seeding the block-wide scan.
+    __device__ float operator()(float block_aggregate)
     {
-        BLOCK_THREADS = BLOCK_THREADS_
-    };
-    enum
-    {
-        ITEMS_INCREMENT = 32
-    };
-    // enum { KEYS_PER_LDG = 2 * 4 / sizeof(Key_Data_Type_) };
-    enum
-    {
-        KEYS_PER_LDG = KEYS_PER_LDG_
-    };
+        float old_prefix = running_total;
+        running_total += block_aggregate;
+        return old_prefix;
+    }
 };
 
-struct TopKPerSegmentContext {
-    TopKPerSegmentContext(): sm_count(0), sm_shared_size(0), sm_version(0){};
-    int sm_count;
-    int sm_shared_size;
-    int sm_version;
+struct TopPSortParams {
+    void*  workspace{};
+    size_t workspace_size{};
```

Comment:
some members use in-place initialization while some don't. 
Any special reason for doing that?